File: backend/src/services/documentParser.ts
import pdf from 'pdf-parse';
import mammoth from 'mammoth';
import fs from 'fs/promises';

export async function parsePDF(filePath: string): Promise<string> {
  const dataBuffer = await fs.readFile(filePath);
  const data = await pdf(dataBuffer);
  return data.text;
}

export async function parseWord(filePath: string): Promise<string> {
  const result = await mammoth.extractRawText({ path: filePath });
  return result.value;
}

export async function parseDocument(filePath: string, fileType: string): Promise<string> {
  const extension = fileType.toLowerCase();

  if (extension === 'pdf' || extension === 'application/pdf') {
    return parsePDF(filePath);
  } else if (
    extension === 'docx' ||
    extension === 'doc' ||
    extension === 'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
  ) {
    return parseWord(filePath);
  } else if (extension === 'txt' || extension === 'text/plain') {
    return fs.readFile(filePath, 'utf-8');
  }

  throw new Error(`Unsupported file type: ${fileType}`);
}

File: backend/app/services/document_parser.py
import PyPDF2
from docx import Document as DocxDocument

def parse_pdf(file_path: str) -> str:
    """Extract text from PDF file"""
    text = ""
    with open(file_path, 'rb') as file:
        pdf_reader = PyPDF2.PdfReader(file)
        for page in pdf_reader.pages:
            text += page.extract_text()
    return text

def parse_docx(file_path: str) -> str:
    """Extract text from Word document"""
    doc = DocxDocument(file_path)
    text = "\n".join([paragraph.text for paragraph in doc.paragraphs])
    return text

def parse_text(file_path: str) -> str:
    """Read plain text file"""
    with open(file_path, 'r', encoding='utf-8') as file:
        return file.read()

def parse_document(file_path: str, file_type: str) -> str:
    """Parse document based on file type"""
    file_type = file_type.lower()

    if 'pdf' in file_type:
        return parse_pdf(file_path)
    elif 'word' in file_type or 'docx' in file_type or 'doc' in file_type:
        return parse_docx(file_path)
    elif 'text' in file_type or 'txt' in file_type or 'plain' in file_type:
        return parse_text(file_path)
    else:
        raise ValueError(f"Unsupported file type: {file_type}")
